### DISCUSSION TITLE
[Support] PiVPN with Unbound and PiHole - no DNS

### DISCUSSION THREAD

**Discussion by nrampage** [pivpn/pivpn#1476] [Support]
## In raising this issue, I confirm the following:

`{please fill the checkboxes, e.g: [X]}`

- [X] I have read the [documentation](https://docs.pivpn.io)
- [X] I have read and understood the [contributors guide](https://github.com/pivpn/pivpn/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting can be *is* directly related to the pivpn installer script.

### Have you searched for similar issues and solutions?
    (yes)

### Has your install failed? 

```
no
```

#### What type of hardware are you running PiVPN at?

```
Raspberrypi 3b
```
#### Output of `uname -a`

```
 Linux raspberrypi 4.19.66-v7+ #1253 SMP Thu Aug 15 11:49:46 BST 2019 armv7l GNU/Linux
```

#### Output of `cat /etc/os-release`

```
PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
NAME="Raspbian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

### If install failed Please provide the console output of  `curl -L https://install.pivpn.io | bash`

```
not applicable
```

### Issue

I am attempting to use Unbound with PiHole and PiVPN.

Unbound works for Pihole and DNS on my LAN (192.168.1.0/24) (Pihole is configured as DNS server on 192.168.1.200). I can connect fine with Wireguard as well, but I can't get DNS to resolve unless I change the configuration for Wireguard clients to use a public IP for DNS, like 1.1.1.1 or 9.9.9.9.

My setup followed information I found at [https://github.com/notasausage/pi-hole-unbound-wireguard](https://github.com/notasausage/pi-hole-unbound-wireguard). Since that wasn't working 100% correctly, I consulted [https://docs.pi-hole.net/guides/vpn/wireguard/server/](https://docs.pi-hole.net/guides/vpn/wireguard/server/) and didn't get any further.

I've gone so far as to reset all my iptables configuration and start over from scratch using the suggested Pihole rule set from their own documentation, then running PIVPN -d and allowing it to create forwarding rules there. I am at a loss, and would appreciate any input as to what I can look at to find the misconfiguration.

I have attempting monitoring tcpdump for traffic on port 53 and 5335 (unbound is configured for 5335), and don't see any when I have a VPN client connected, but I definitely could be doing something wrong with tcpdump.

### Console output of      `pivpn debug`

```
::::            PiVPN debug              ::::
=============================================
::::            Latest commit            ::::
Branch: master
Commit: f80b0a7962d91862132c0a4abd65c1e67bd37bd7
Author: 4s3ti
Date: Fri Dec 3 21:41:13 2021 +0100
Summary: Update Changelog
=============================================
::::        Installation settings        ::::
PLAT=Raspbian
OSCN=stretch
USING_UFW=0
IPv4dev=eth0
dhcpReserv=1
IPv4addr=192.168.1.200/24
IPv4gw=192.168.1.1
install_user=pi
install_home=/home/pi
VPN=wireguard
pivpnPORT=51820
pivpnDNS1=192.168.1.200
pivpnDNS2=
pivpnHOST=REDACTED
INPUT_CHAIN_EDITED=1
FORWARD_CHAIN_EDITED=1
pivpnPROTO=udp
pivpnMTU=1420
pivpnDEV=wg0
pivpnNET=10.6.0.0
subnetClass=24
ALLOWED_IPS="0.0.0.0/0, ::0/0"
UNATTUPG=1
INSTALLED_PACKAGES=()
=============================================
::::  Server configuration shown below   ::::
[Interface]
PrivateKey = server_priv
Address = 10.6.0.1/24
MTU = 1420
ListenPort = 51820
DNS = 192.168.1.200

# Replace eth0 with the interface open to the internet (e.g might be wlan0 if wifi)
PostUp = iptables -A FORWARD -i %i -j ACCEPT; iptables -A FORWARD -o %i -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
PostDown = iptables -D FORWARD -i %i -j ACCEPT; iptables -D FORWARD -o %i -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE

### begin client1 ###
[Peer]
PublicKey = client1_pub
PresharedKey = client1_psk
AllowedIPs = 10.6.0.2/32
### end client1_iPhone ###
=============================================
::::  Client configuration shown below   ::::
[Interface]
PrivateKey = client1_priv
Address = 10.6.0.2/24
DNS = 10.6.0.1

[Peer]
PublicKey = server_pub
PresharedKey = client1_psk
Endpoint = REDACTED:51820
AllowedIPs = 0.0.0.0/0, ::0/0
=============================================
::::    Recursive list of files in       ::::
::::    /etc/wireguard shown below       ::::
/etc/wireguard:
backup
configs
keys
wg0.conf
wg0.conf.save
wg0.conf.save.1

/etc/wireguard/backup:
configs
keys
wg0.conf
wg0.conf.save
wg0.conf.save.1

/etc/wireguard/backup/configs:
<file names removed for privacy>

/etc/wireguard/backup/keys:
<key names removed for privacy>

/etc/wireguard/configs:
clients.txt
client1.conf

/etc/wireguard/keys:
client1_priv
client1_psk
client1_pub
server_priv
server_pub
=============================================
::::            Self check               ::::
:: [OK] IP forwarding is enabled
:: [OK] Iptables MASQUERADE rule set
:: [OK] Iptables INPUT rule set
:: [OK] Iptables FORWARD rule set
:: [OK] WireGuard is running
:: [OK] WireGuard is enabled (it will automatically start on reboot)
:: [OK] WireGuard is listening on port 51820/udp
=============================================
:::: Having trouble connecting? Take a look at the FAQ:
:::: https://docs.pivpn.io/faq
=============================================
:::: WARNING: This script should have automatically masked sensitive       ::::
:::: information, however, still make sure that PrivateKey, PublicKey      ::::
:::: and PresharedKey are masked before reporting an issue. An example key ::::
:::: that you should NOT see in this log looks like this:                  ::::
:::: YIAoJVsdIeyvXfGGDDadHh6AxsMRymZTnnzZoAb9cxRe                          ::::
=============================================
::::            Debug complete           ::::
:::
::: Debug output completed above.
::: Copy saved to /tmp/debug.log
:::
```

### Have you taken any steps towards solving your issue?

See above! 

Also posted to Discussions, which I'm going to remove now (I'm new to github and didn't know the best place to raise this question initially).

EDIT: corrected hyperlinks in "Issue" section

**Comment by TinCanTech** [comment]
FYI: Using common subnets, like 192.168.1.0/24, for your server LAN, is very likely to cause routing conflicts with your clients. Use something more unique  from RFC1918 ranges.

FTR: It is also possible to over-ride routing conflicts by using: `--redirect-gateway def1 block-local`, which routes the CLIENT lan into the VPN, except for the client default-gateway.

**Comment by nrampage** [comment]
Note it works as desired as long as I use a public/external DNS.  My Wireguard server IP is 10.6.0.1, so I'm not sure I completely understand your point...most likely my ignorance on the issue.

Also, apologies for the giant text in the OP, I can't figure out where the formatting issues are in my post.  Will gladly correct if anyone knows why.

EDIT:  I should note PiHole is listening for Unbound on 127.0.0.1#5335.

**Comment by TinCanTech** [comment]
You are using wireguard, please ignore my previous comment, which pertains to OpenVPN.

**Comment by coolapso** [comment]
Updated your issue description and fixed your formatting issues ... it was terribly painful to look at :p 

Moving to discussions as its the correct place for support requests.

**Reply by coolapso** [reply]
I have also deleted the discussion you have opened previously to avoid confusion.

**Reply by nrampage** [reply]
> Updated your issue description and fixed your formatting issues ... it was terribly painful to look at :p
> 
> Moving to discussions as its the correct place for support requests.

Yes, thank you. :)  Any chance you know where I went wrong in my pasting of information?

**Reply by coolapso** [reply]
some backticks ` were misplaced :) 

Edit: there's always the "preview" tab to try to figure it out :D

**Comment by nrampage** [comment]
I found the answer, finally!

https://www.reddit.com/r/WireGuard/comments/cd8ouo/wireguard_not_using_pihole_for_dns/ gave me the idea to edit the /etc/dnsmasq.d/02-pivpn.conf file to add

`interface=wg0`

I did that, restarted Wireguard via `wgquick down wg0` followed by `wgquick up wg0`, then ran `pihole restartdns`.  No idea if all those steps were necessary.

Now my Wireguard connections are resolving DNS.  Also confirmed no "leaks" via dnsleak.com.
